hearts/game/manager.py

```diff
@@ -176,6 +194,130 @@ def get_bot_card_to_play(self, bot: Player) -> Card:
+        synchronous, we must split passing into two steps:
+        1. Each player signifies which cards they are going to pass.
+        2. We pass all cards at the same time.
+
+        This method is the first step and the `pass_cards` method is the second
+        step.
+        """
+        # Validate the current deal hasn't passed yet.
+        if self.current_deal.has_passed:
+            return
+
+        # Check if passing is necessary. On every 4th round we skip passing.
+        direction = self.get_pass_direction()
+        if direction is None:
+            self.current_deal.has_passed = True
+            self.current_deal.save()
+            return
+
+        # Validate the correct number of cards are getting passed.
+        if len(cards) != 3:
+            return
+
+        for card in cards:
+            card.to_pass = True
+            card.save()
+
+    def pass_cards(self) -> None:
+        """
+        Pass cards to adjacent players.
+
+        This happens at the start of the round. Right now bots will pass three
+        random cards.
+        """
+        deal = self.current_deal
+        direction = self.get_pass_direction()
+        if direction is None:
+            deal.has_passed = True
+            deal.save()
+            logger.info('Skipping passing on 4th deal')
+            return
+
+        for player in self.game.players:
+            cards_to_pass = Card.objects.filter(
+                player=player,
+                deal=deal,
+                to_pass=True,
+            ).count()
+            if cards_to_pass != 3:
+                if not player.bot:
+                    logger.info(f'Waiting for player {self.game.get_player_index(player.id)} to pass')
+                    return
+                cards = list(Card.objects.filter(
+                    player=player,
+                    deal=deal,
+                    to_pass=False,
+                ).order_by('?')[:3])
+                for card in cards:
+                    card.to_pass = True
+                    card.save()
+
+        pass_map = {
+            self.game.player_1_id: {
+                'left': self.game.player_2,
+                'right': self.game.player_4,
+                'top': self.game.player_3,
+            },
+            self.game.player_2_id: {
+                'left': self.game.player_3,
+                'right': self.game.player_1,
+                'top': self.game.player_4,
+            },
+            self.game.player_3_id: {
+                'left': self.game.player_4,
+                'right': self.game.player_2,
+                'top': self.game.player_1,
+            },
+            self.game.player_4_id: {
+                'left': self.game.player_1,
+                'right': self.game.player_3,
+                'top': self.game.player_2,
+            },
+        }
+
+        for player in self.game.players:
+            receiving_player = pass_map[player.id][direction]
+            Card.objects.filter(
```

Comment:
287-294?

Reply:
Just a heads up, in GitHub you can select a range of lines to leave a comment on by clicking on the starting line and dragging down.

Anyway, the query in question is this:
```python
Card.objects.filter(
    deal=self.current_deal,
    player=player,
    to_pass=True,
).update(
    player=receiving_player,
    to_pass=False,
)
```
This query is finding all cards this player has marked to pass and re-assigning ownership to the player who is receiving the pass.

The SQL would be something like this:
```sql
UPDATE
  Card
SET
  player = receiving_player,
  to_pass = False
WHERE
  deal = current_deal AND player = player AND to_pass = True;
```